net/dispatcher.go

```diff
@@ -19,44 +23,158 @@ import (
 	"github.com/frain-dev/convoy/util"
 )
 
+var (
+	ErrAllowListIsRequired = errors.New("allowlist is required")
+	ErrBlockListIsRequired = errors.New("blocklist is required")
+	ErrLoggerIsRequired    = errors.New("logger is required")
+	ErrInvalidIPPrefix     = errors.New("invalid IP prefix")
+)
+
+type DispatcherOption func(d *Dispatcher) error
+
 type Dispatcher struct {
-	client *http.Client
+	logger    *log.Logger
+	transport *http.Transport
+	client    *http.Client
+	rules     *netjail.Rules
 }
 
-func NewDispatcher(httpProxy string, licenser license.Licenser, enforceSecure bool) (*Dispatcher, error) {
-	d := &Dispatcher{client: &http.Client{}}
+// ProxyOption defines an HTTP proxy which the client will use. It fails-open the string isn't a valid HTTP URL
+func ProxyOption(l license.Licenser, httpProxy string) DispatcherOption {
+	return func(d *Dispatcher) error {
+		if httpProxy == "" {
+			return nil
+		}
+
+		if l.UseForwardProxy() {
+			proxyUrl, isValid, err := d.setProxy(httpProxy)
+			if err != nil {
+				return err
+			}
 
-	tr := &http.Transport{
-		MaxIdleConns:          100,
-		IdleConnTimeout:       10 * time.Second,
-		MaxIdleConnsPerHost:   10,
-		TLSHandshakeTimeout:   3 * time.Second,
-		ExpectContinueTimeout: 1 * time.Second,
+			if isValid {
+				d.transport.Proxy = http.ProxyURL(proxyUrl)
+			}
+		}
+
+		return nil
 	}
+}
 
-	if licenser.UseForwardProxy() {
-		proxyUrl, isValid, err := d.setProxy(httpProxy)
-		if err != nil {
+// AllowListOption sets a list of IP prefixes which will outgoing traffic will be granted access
+func AllowListOption(allowList []string) DispatcherOption {
+	return func(d *Dispatcher) error {
+		if len(allowList) == 0 {
```

Comment:
@jirevwe Is there a reason these values can't be empty? Making it required can be a proactive measure to keep our users safe. But this will break a lot of prod deployments.